### DISCUSSION TITLE
NootRX with an eGPU (Works!)

### DISCUSSION THREAD

**Discussion by ivan-ef** [ChefKissInc/NootRX#14] [Show and tell]
Will NootRX work with external thunderbolt GPU box?

**Comment by ExtremeXT** [comment]
@ivan-ef No idea, try it and let us know! But your Thunderbolt will obviously have to be setup properly.

**Comment by ivan-ef** [comment]
> @ivan-ef No idea, try it and let us know! But your Thunderbolt will obviously have to be setup properly.

My thunderbolt egpu is working correctly with Vega 56. I want to try rx6700xt
I have whatevergreen kext so how can I avoid conflict?

**Comment by ExtremeXT** [comment]
> > @ivan-ef No idea, try it and let us know! But your Thunderbolt will obviously have to be setup properly.
> 
> My thunderbolt egpu is working correctly with Vega 56. I want to try rx6700xt I have whatevergreen kext so how can I avoid conflict?

By removing it?

**Reply by ivan-ef** [reply]
> By removing it?

Now it's working over thunderbolt EGPU. 
key moment was to make OC Clean Snapshot in ProperTree.

<img width="240" alt="Снимок экрана 2023-11-19 в 19 12 06" src="https://github.com/ChefKissInc/NootRX/assets/67600440/ffe0b33f-440e-4854-a605-4ec88901112f">
<img width="240" src="https://github.com/ChefKissInc/NootRX/assets/67600440/43d95f88-6eb3-465a-8d53-db6634db43af">
<img width="240" src="https://github.com/ChefKissInc/NootRX/assets/67600440/504d6d50-e5c6-4fab-a232-2f5f3d43c20b">

**Comment by ExtremeXT** [comment]
> > By removing it?
> 
> Now it's working over thunderbolt EGPU. key moment was to make OC Clean Snapshot in ProperTree.
> 
> <img alt="Снимок экрана 2023-11-19 в 19 12 06" width="240" src="https://user-images.githubusercontent.com/67600440/284077552-ffe0b33f-440e-4854-a605-4ec88901112f.png"> <img alt="" width="240" src="https://user-images.githubusercontent.com/67600440/284076268-43d95f88-6eb3-465a-8d53-db6634db43af.jpg"> <img alt="" width="240" src="https://user-images.githubusercontent.com/67600440/284076286-504d6d50-e5c6-4fab-a232-2f5f3d43c20b.jpg">

Great! You're welcome!

**Comment by ivan-ef** [comment]
> Great! You're welcome!

However, most apps just stuck when using this GPU... Don't know how to debug. Other natively supported GPU works fine

**Comment by ExtremeXT** [comment]
> > Great! You're welcome!
> 
> However, most apps just stuck when using this GPU... Don't know how to debug. Other natively supported GPU works fine

See if you have any .gpuRestart file in /Library/Logs/DiagnosticReports.

**Comment by ivan-ef** [comment]
> See if you have any .gpuRestart file in /Library/Logs/DiagnosticReports.

Looks like I don't have it.
Maybe a diag files from apps that stuck will help... 

[app.zip](https://github.com/ChefKissInc/NootRX/files/13443296/app.zip)

**Comment by Maheshroy50** [comment]
Does this work with Rx6400 GPU.

**Comment by Maheshroy50** [comment]
#include <IOKit/IOService.h>
#include <IOKit/IOLib.h>

class RX6400Kext : public IOService {
    OSDeclareDefaultStructors(RX6400Kext)

public:
    virtual bool init(OSDictionary* dictionary = nullptr) override;
    virtual void free() override;
    virtual IOService* probe(IOService* provider, SInt32* score) override;
    virtual bool start(IOService* provider) override;
    virtual void stop(IOService* provider) override;

private:
    // Your private members and methods go here
};

// Macros for the kernel module loading
OSDefineMetaClassAndStructors(RX6400Kext, IOService)

bool RX6400Kext::init(OSDictionary* dictionary) {
    if (!super::init(dictionary)) {
        return false;
    }

    // Your initialization code goes here

    return true;
}

void RX6400Kext::free() {
    // Your cleanup code goes here

    super::free();
}

IOService* RX6400Kext::probe(IOService* provider, SInt32* score) {
    IOService* result = super::probe(provider, score);

    // Your probing code goes here

    return result;
}

bool RX6400Kext::start(IOService* provider) {
    if (!super::start(provider)) {
        return false;
    }

    // Your startup code goes here

    return true;
}

void RX6400Kext::stop(IOService* provider) {
    // Your shutdown code goes here

    super::stop(provider);
}
is this the formate of the code. i just try to do something is this correct formate.

**Comment by Maheshroy50** [comment]
// Inside the RX6400Driver class definition

private:
    // ... (other members and methods)

    // Example GPU-related members
    uint32_t computeUnits;
    uint32_t boostFrequency;
    uint32_t gameFrequency;
    uint32_t rayAccelerators;
    float peakPixelFillRate;
    float peakTextureFillRate;
    float peakHalfPrecisionComputePerformance;
    float peakSinglePrecisionComputePerformance;
    uint32_t ROPs;
    uint32_t streamProcessors;
    uint32_t textureUnits;
    uint32_t transistorCount;

    // Example power-related members
    uint32_t typicalBoardPower;
    uint32_t minimumPSURecommendation;

    // Example GPU memory-related members
    uint32_t infinityCacheSize;
    uint32_t memorySpeed;
    uint32_t maxMemorySize;
    const char* memoryType;
    uint32_t memoryInterface;
    float memoryBandwidth;

    // Example rendering format-related members
    bool supportsHDMI4K;
    bool supports4KH264Decode;
    bool supports4KH264Encode;
    bool supportsH265HEVCDecode;
    bool supportsH265HEVCEncode;
    bool supportsAV1Decode;

    // Example connectivity-related members
    uint32_t displayPortVersion;
    bool supportsHDMI21VRRAndFRL;
    bool supportsUSBTypeC;

    // Example dimensions-related members
    const char* slotSize;

    // Example methods to query GPU specifications
    void queryGPUSpecifications();
    void queryPowerSpecifications();
    void queryGPUMemorySpecifications();
    void queryRenderingFormatSpecifications();
    void queryConnectivitySpecifications();

    // ... (other members and methods)
does this is work

**Comment by Maheshroy50** [comment]
// Inside the RX6400Driver class implementation

void RX6400Driver::queryGPUSpecifications() {
    // Example: Set GPU-related specifications
    computeUnits = 12;
    boostFrequency = 2321;
    gameFrequency = 2039;
    rayAccelerators = 12;
    peakPixelFillRate = 74.3;  // GP/s
    peakTextureFillRate = 111.4;  // GT/s
    peakHalfPrecisionComputePerformance = 7.13;  // TFLOPs
    peakSinglePrecisionComputePerformance = 3.57;  // TFLOPs
    ROPs = 32;
    streamProcessors = 768;
    textureUnits = 48;
    transistorCount = 5.4;  // B
}

void RX6400Driver::queryPowerSpecifications() {
    // Example: Set power-related specifications
    typicalBoardPower = 53;  // W
    minimumPSURecommendation = 350;  // W
}

void RX6400Driver::queryGPUMemorySpecifications() {
    // Example: Set GPU memory-related specifications
    infinityCacheSize = 16;  // MB
    memorySpeed = 16;  // Gbps
    maxMemorySize = 4;  // GB
    memoryType = "GDDR6";
    memoryInterface = 64;  // bit
    memoryBandwidth = 128;  // GB/s
}

void RX6400Driver::queryRenderingFormatSpecifications() {
    // Example: Set rendering format-related specifications
    supportsHDMI4K = true;
    supports4KH264Decode = true;
    supports4KH264Encode = false;
    supportsH265HEVCDecode = true;
    supportsH265HEVCEncode = false;
    supportsAV1Decode = false;
}

void RX6400Driver::queryConnectivitySpecifications() {
    // Example: Set connectivity-related specifications
    displayPortVersion = 1.4;  // Version 1.4a
    supportsHDMI21VRRAndFRL = true;
    supportsUSBTypeC = false;
}

// ... (other methods)

**Comment by qingquan119** [comment]
纳维24 什么时候能支持？？
